streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -595,6 +595,72 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
       ZStream.fromQueue(queue).unTake
     }
 
+  /**
+   * Creates a stream that passes its elements through a queue that drops elements
+   * (dropping or sliding). It implements a promise that signals abrupt completion
+   * of the stream due to an error or the end of the stream. After the promise has
+   * been completed, it continues to offer the current (at the time) contents of the
+   * queue, until it has been exhausted, at which point it finishes with the signal
+   * from the promise.
+   */
+  private final def bufferSignal[E1 >: E, A1 >: A](queue: Queue[Take[E1, A1]]): ZStream[R, E1, A1] =
+    ZStream.managed {
+      for {
+        signal <- Promise.make[Cause[E], Unit].toManaged_
```

Comment:
The `Cause[E]` here is basically redundant because you can always recover it by using `Promise#halt` from the producer side and using `await.sandbox` or `await.run` from the consumer side

Reply:
Thanks, I'll keep an eye out next time.

This has become irrelevant for the new implementation.